File: python/91.decode-ways.py
#
# @lc app=leetcode id=91 lang=python3
#
# [91] Decode Ways
#

# @lc code=start
def countNumDecodings(s: str) -> int:
    if len(s) == 0:
        return 1
    # '0' or 'x'
    if len(s) == 1:
        if s[0] == '0':
            return 0
        return 1
    other_way = 0
    # Count for 'x'
    if s[0] == '0':
        return 0
    else:
        if int(s[0:2]) <= 26:
            other_way = countNumDecodings(s[2:])
        return countNumDecodings(s[1:]) + other_way
class Solution:
    def numDecodings(self, s: str) -> int:
        return countNumDecodings(s)
# @lc code=end

if __name__ == "__main__":
    a = Solution()
    print("result: ", a.numDecodings("9317949759856497357254398763219839323723136763131916377913495416692666785978758414629119614215967159"))

File: golang/91.decode-ways.go
package main

import (
	"fmt"
	"strconv"
)

/*
 * @lc app=leetcode id=91 lang=golang
 *
 * [91] Decode Ways
 */
// @lc code=start
func numDecodings(s string) int {
	if len(s) == 0 {
		return 1
	}
	// '0' or 'x'
	if len(s) == 1 {
		if s[0] == '0' {
			return 0
		}
		return 1
	}
	otherWay := 0
	if s[0] == '0' {
		return 0
	} else {
		val, err := strconv.ParseInt(s[:2], 10, 32)
		if err == nil && val <= 26 {
			otherWay = numDecodings(s[2:])
		}
		return numDecodings(s[1:]) + otherWay
	}
}

// @lc code=end

func main() {
	fmt.Println("test: ", numDecodings("9317949759856497357254398763219839323723136763131916377913495416692666785978758414629119614215967159"))
}
